clippy_lints/src/doc.rs

```diff
@@ -512,24 +516,45 @@ fn check_doc<'a, Events: Iterator<Item = (pulldown_cmark::Event<'a>, Range<usize
             End(Link(..)) => in_link = None,
             Start(Heading(_)) => in_heading = true,
             End(Heading(_)) => in_heading = false,
+            Start(Paragraph) => {
+                ticks_unbalanced = false;
+                let (_, span) = get_current_span(spans, range.start);
+                paragraph_span = first_line_of_span(cx, span);
+            },
+            End(Paragraph) => {
+                if ticks_unbalanced {
+                    span_lint(
+                        cx,
+                        DOC_MARKDOWN,
+                        paragraph_span,
+                        "backticks are unbalanced; one may be missing a pair",
```

Comment:
This message seems a bit off to me in comparison to other messages, but I can't quite place why. @flip1995 could you give this a second look?

Reply:
If there is punctuation in between the message, you usually want to split it up in error message + help message. Which you can do here with `span_lint_and_help`. So put the first part in the error message and the second part in the help message.

Bonus points when you also provide the span to the unmatched backtick as the span for the help message.

Reply:
Splitting the message up into `span_lint_and_help` would make sense to me as well :+1: 

> Bonus points when you also provide the span to the unmatched backtick as the span for the help message.

It is sadly a bit difficult to automatically determine which backtick is unmatched in longer paragraphs with multiple snippets. It could maybe show the backtick if there is only one?

Example:
```rust
/// Hello World! `some-code-here` more
/// text in the next line with an unmatched `backtick
/// and finally `the-end`
```

How should the lint know that it's the middle backtick that it should highlight? Marking the whole paragraph seems elegant enough to me.

Reply:
Yeah, that's why I said bonus points. If @ebobrow has a good idea how to do this, that would be nice, but if not, just pointing to the paragraph is definitely good enough.

Reply:
Agreed. I don't think it's possible.

For the error message, does something like this look good?

```
error: backticks are unbalanced
  --> $DIR/unbalanced_ticks.rs:7:1
   |
LL | / /// This is a doc comment with `unbalanced_tick marks and several words that
LL | | /// should be `encompassed_by` tick marks because they `contain_underscores`.
LL | | /// Because of the initial `unbalanced_tick` pair, the error message is
LL | | /// very `confusing_and_misleading`.
   | |____________________________________^
   |
   = note: `-D clippy::doc-markdown` implied by `-D warnings`
   = help: a backtick may be missing a pair
```

Reply:
I think that looks good 👍